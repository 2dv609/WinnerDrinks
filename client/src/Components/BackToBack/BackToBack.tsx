import { useState, useEffect } from 'react';
import GameProps from '../GameProps';
import '../../App.css';

const NUM_OF_PLAYERS = 2; // How many players do you need? 

function BackToBack(props: any) {
    const gp: GameProps = props.gp; // Contains the methods done and getPlayers. 
    const [players, setPlayers] = useState(gp.getPlayers(NUM_OF_PLAYERS)); // players is an array 

    useEffect(() => { 
        return () => { // Return a function for code cleanup. This will set new players 

            setPlayers(gp.getPlayers(NUM_OF_PLAYERS));
        }
    }, [gp])

    // Either both win, or both lose.
    return (
        <div className="box container">
            <div className="content">{props.gameEvent.question}</div>
            <h3 className="title is-6">Were both correct?</h3>

            <button className="button" onClick={() => {
                gp.addScore(players[0], 1)
                gp.addScore(players[1], 1)
                gp.makeWinnerAlert(players)
                gp.chooseRandomNewGame()
            }}>{'They were correct!'}</button>
            <button className="button" onClick={() => {
                gp.makeWinnerAlert(null)
                gp.chooseRandomNewGame()
            }}>{'They were wrong!'}</button>
        </div>
    )
};

export default BackToBack;