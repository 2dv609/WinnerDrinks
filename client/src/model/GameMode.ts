export enum GameMode {
    STANDARD, HIGHSCORE
}