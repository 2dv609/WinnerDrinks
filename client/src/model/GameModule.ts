export interface IGameModuleSetting {
    name: string,
    active: boolean,
    enable: boolean,
    index: number
}

/**
 * Class containing values for implemented game modules in the application
 */
export default class GameModule {
    readonly TRIVIA = 'trivia'
    readonly PARTY = 'party'
    readonly BACK_TO_BACK = 'back-to-back'

    /**
     * Get game module names to use for api calls and indexeDB tables.
     *
     * @returns {string[]} - Array of string for api calls and indexedDB tables.
     */
    public getNames(): string[] {
      return [
        this.TRIVIA,
        this.PARTY,
        this.BACK_TO_BACK
      ]
    }

    /**
     * Get default game module settings.
     *
     * @returns {IGameModuleSetting[]} - An array of game module settings.
     */
    public getSettings(): IGameModuleSetting[] {
      return [
        { name: 'Wheel', enable: true, active: true, index: 0 },
        { name: 'Party', enable: true, active: true, index: 1 },
        { name: 'BackToBack', enable: false, active: false, index: 2 },
        { name: 'Trivia', enable: true, active: true, index: 3 },
      ]
    }
}
