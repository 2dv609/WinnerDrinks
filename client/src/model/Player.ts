/**
 * A player class for keeping track of the player names,
 * their score and whether they are currently active in the game or not.
 * @author Delfi Sehidic
 */

class Player {
    private _name: string = ''; // The name of the player as registered in the input
    isActive: boolean; // Is the player currently playing or paused for e.g. a bathroom visit?
    score: number;

    /**
     * Constructs a new player object.
     * @param name Must be at least 3 characters long and not exceed 10.
     */
    constructor(name: string) {
      this.name = name;
      this.isActive = true;
      this.score = 0;
    }
    /**
     * Overwrites the old score of the player and sets it to newScore.
     * @param newScore
     */
    setScore(newScore: number) {
      this.score = newScore;
    }
    /**
     * Updates the score. For example, if the score is 10 and points is 1, the score will now be 11.
     *
     * @param points Added to the existing score. Can be negative.
     */
    addScore(points: number) {
      this.score += points;
    }
    /**
     * Returns a string representation of the player.
     * @param debug If set to true, it also returns the score and activity of the player.
     * @returns Player name or full player stats.
     */
    toString(debug?:boolean): string {
      if (debug) {
        return `Name: ${this._name} Score: ${this.score} isActive: ${this.isActive}`;
      }
      return this._name;
    }

    /**
     * Setter for property _name.
     */
    set name(name: string) {
      if (name.length < 1 || name.length > 10) {
        throw Error('Name must be at least 1 characters and not exceed 10.');
      }
      if (name.split(' ').length > 1) {
        throw Error('Name can not contain any spaces.');
      }
      this._name = name;
    }

    /**
     * Getter for property _name.
     */
    get name() {
      return this._name
    }
}

export default Player;
