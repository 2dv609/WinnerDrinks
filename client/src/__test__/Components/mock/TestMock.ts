import { GameModuleProps } from '../../../Components/GameModuleProps'
import Player from '../../../model/Player'

export const playersMock: Player[] = [new Player('Tom'), new Player('Carl'), new Player('Sandra')]

const makeWinnerAlertMock = (p: Player | Player[] | null, message?: string): void => {}
const chooseRandomNewGameMock = (): void => {}
const removeGameEventMock = (gameEventId: string, gameEvents: GameEventAPI): void => {}
const getPlayersMock = (amount: number, players: Player[]): Player[] => {return players}
const addScoreMock = (p: Player) => void {}


export const gameServiceMock: GameModuleProps = {
    players: playersMock,
    getPlayers: getPlayersMock,
    addScore: addScoreMock,
    makeWinnerAlert: makeWinnerAlertMock,
    chooseRandomNewGame: chooseRandomNewGameMock,
    removeGameEvent: removeGameEventMock        
}
