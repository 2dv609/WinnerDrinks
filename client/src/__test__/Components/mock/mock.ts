import Player from '../../../Player'


// Mock GameProps
const addScoreMock = (p: Player, score: number) => {}
const makeWinnerAlertMock = (p: any) => {}
const chooseRandomNewGameMock = () => {}
const getPlayersMock = (amount: number): Player[] => {
    const player1 = new Player('Mats')
    const player2 = new Player('Tom')
    return [player1, player2];
};

export const gamePropsMock = {
    getPlayers: getPlayersMock, // Let the component decide how many players it needs. 
    addScore: addScoreMock, // Adds a score to a player
    makeWinnerAlert: makeWinnerAlertMock, // makes an Alert that corresponds to if there are one, multiple, or no winners
    chooseRandomNewGame: chooseRandomNewGameMock, // chooses a random new game
}

// Mock game event mock
export const gameEventMock = {
    _id: 111,
    question: 'This is a mock question'
}