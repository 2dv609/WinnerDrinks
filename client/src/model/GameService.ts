import Player from './Player'
import { TextGameModuleProps, AnimationGameModuleProps } from '../Components/GameModuleProps'

export default class GameService {

    public addScore = (p: Player, score: number) => {
        p.addScore(score)
    }

    public getNewGameIndex = (currentGameIndex: number, gameModules: (React.FC<TextGameModuleProps> | React.FC<AnimationGameModuleProps>)[], activeGames: any[]): number => {
        let newIndex = currentGameIndex
        while (newIndex === currentGameIndex || !activeGames[newIndex].active) { // Don't allow the same game twice in a row. 
          console.log('while')
          newIndex = Math.floor(Math.random() * gameModules.length)
        }
        return newIndex
    }
    /**
     * 
     * @param nrOfPlayers The amount of players you need for the component
     * @param players An array of players that the function will fetch from. 
     * @returns A sublist from your array that contains the requested amount of active players. 
     */
    public getPlayers = (nrOfPlayers: number, players: Player[]): Player[] => {
        const activePlayers: Player[] = [];
        
        for (let i = 0; i < players.length; i++) {
            if (players[i].isActive) {
                activePlayers.push(players[i])
            }
            
        }

        this.shuffle(activePlayers);
        
        return activePlayers.slice(0, nrOfPlayers);
    }

    private shuffle(array: Player[]) {
        var m = array.length, t, i;
        while (m) {
            i = Math.floor(Math.random() * m--);
            t = array[m];
            array[m] = array[i];
            array[i] = t;
        }
        return array;
    }

    public getRandomGameEvent = (gameEventAPI: GameEventAPI): IBackToBack | IParty | ITrivia => {
        return gameEventAPI.questions[Math.floor(Math.random() * gameEventAPI.questions.length)]
    }

    public getNumActivePlayers = (players: Player[]): number => {
        let numOfActivePlayers = 0

        players.forEach(player => {
            if(player.isActive) numOfActivePlayers++  
        })

        return numOfActivePlayers
      }
}