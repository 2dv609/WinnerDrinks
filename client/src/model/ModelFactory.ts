import GameModuleService from './GameModuleService'
import IGameModuleService from './IGameModuleService'
import GameService from './GameService'
import { IGameModuleSetting } from '../Components/Menu/Navbar'

/**
 * Factory method that return a promise of an instance 
 * of a type IGameModuleSerice with loaded game events or
 * undefined if GameModules coulde not be loaded.
 * 
 * @returns {Promis<IUtilservice | undefined>}
 */
export async function getGameModuleService(): Promise<IGameModuleService | undefined> {

    const gameModuleSerivce: IGameModuleService = new GameModuleService()
    const gameEventIsLoaded: boolean = await gameModuleSerivce.loadGameEvents()

    if (!gameEventIsLoaded) {
        return
    }

    return gameModuleSerivce
}

/**
 * Factory method that return  an instance 
 * of a type IGameService.
 * 
 * @returns {GameService}
 */
 export function getGameService(): GameService {
    return new GameService()
}

export function getGameModuleSettings(): IGameModuleSetting[] {
    
    return [
        { name: 'Wheel', enable: true, active: true, index: 0 },
        { name: 'Party', enable: true, active: true, index: 1 },
        { name: 'BackToBack', enable: false, active: false, index: 2 },
        { name: 'Trivia', enable: true, active: true, index: 3 },
    ]
}
