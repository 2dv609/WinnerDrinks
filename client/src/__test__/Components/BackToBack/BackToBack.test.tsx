import ReactDOM from 'react-dom'
import BackToBack from '../../../Components/BackToBack/BackToBack'
import { render, fireEvent } from '@testing-library/react' 
import '@testing-library/jest-dom/extend-expect'
import { getGameService } from '../../../model/ModelFactory'
import GameService from '../../../model/GameService'
import Player from '../../../model/Player'
import { playersMock, gameServiceMock } from '../mock/TestMock'
import fs from 'fs-extra'
import { join } from 'path'

/* --------------------------------- */
/* Test cases for game module party. */
/* --------------------------------- */

describe('Test suite for game module back to back', () => {

    /* --------------------------------- */
    /* ----- Test cases setup ---------- */
    /* --------------------------------- */

    type RawBackToBackEvent = {
        question: string
    }

    let backToBackEvents: RawBackToBackEvent[]
    
    const gameService: GameService = getGameService()
    const currentPlayers: Player[] = gameService.getPlayers(2, playersMock)
    const backToBackQuestionMock: IParty | ITrivia | IBackToBack = {
        _id: 'mockID',
        question: 'mockQuestion'
    }

    beforeAll(async () => {
        const backToBackEventsPath: string = join( __dirname, '../../../../../server/data/back-to-back.json')
        backToBackEvents = await fs.readJson(backToBackEventsPath)
    })

    test('Game module back to back should use props gameService, gameEvent and currentPlayers', () => {
        const div = document.createElement('div')
        ReactDOM.render(<BackToBack currentPlayers={currentPlayers} gameService={gameServiceMock} gameEvent={backToBackQuestionMock}></BackToBack>, div)
    })

    test.skip('T1.MBB.S.1: Back to back game events should contain at least 20 game events', () => {
        if (!backToBackEvents) {
            return
        }

        expect(backToBackEvents.length >= 20).toBeTruthy()
    })

    test('T1.MBB.S.2: Back to back game events should contain only unique game events', () => {
        if (!backToBackEvents) {
            return
        }

        const questions: string[] = []
        
        backToBackEvents.forEach((event: RawBackToBackEvent): void => {
            questions.push(event.question)
        })

        expect(questions.length === new Set(questions).size).toBeTruthy()
    })

    test('T1.MBB.S.3: Back to back game events should not have more than 60 words', () => {
        if (!backToBackEvents) {
            return
        }

        backToBackEvents.forEach((event: RawBackToBackEvent ): void => {
            expect(event.question.split(' ').length <= 60).toBeTruthy()
        })
    })


    test('T1.MBB.UI.1: Back to back game events should contain 2 active game participants names', () => {
        const { getByTestId } = render(<BackToBack currentPlayers={currentPlayers} gameService={gameServiceMock} gameEvent={backToBackQuestionMock}></BackToBack>)
        const textContentCurrentPlayers: string | null = getByTestId('current-players').textContent

        expect(isPlayersIncludedInGameEvent(textContentCurrentPlayers, currentPlayers)).toBeTruthy()
    })

    
    test('T1.MBB.UI.2: Game module back to back should display a back to back game event', () => {
        const { getByTestId } = render(<BackToBack currentPlayers={currentPlayers} gameService={gameServiceMock} gameEvent={backToBackQuestionMock}></BackToBack>)
        expect(getByTestId('game-event')).toHaveTextContent(backToBackQuestionMock.question)
        expect(getByTestId('button-correct')).toHaveTextContent('They were correct!')
        expect(getByTestId('button-incorrect')).toHaveTextContent('They were wrong!')
    })

    test('T1.MBB.UI.3: Select back to back events result should trigger transition to next game round, button correct', () => {
        const { getByTestId } = render(<BackToBack currentPlayers={currentPlayers} gameService={gameServiceMock} gameEvent={backToBackQuestionMock}></BackToBack>)
        
        const addScoreSpy = jest.spyOn(gameServiceMock, 'addScore')
        const makeWinnerAlertSpy = jest.spyOn(gameServiceMock, 'makeWinnerAlert')
        const chooseNewRandomGameSpy = jest.spyOn(gameServiceMock, 'chooseRandomNewGame')
        
        fireEvent.click(getByTestId('button-correct'))

        expect(addScoreSpy).toBeCalledTimes(2)
        expect(makeWinnerAlertSpy).toBeCalled()
        expect(chooseNewRandomGameSpy).toBeCalled()
    })

    test('T1.MBB.UI.3: Select back to back events result should trigger transition to next game round, button incorrect', () => {
        const { getByTestId } = render(<BackToBack currentPlayers={currentPlayers} gameService={gameServiceMock} gameEvent={backToBackQuestionMock}></BackToBack>)
        
        const makeWinnerAlertSpy = jest.spyOn(gameServiceMock, 'makeWinnerAlert')
        const chooseNewRandomGameSpy = jest.spyOn(gameServiceMock, 'chooseRandomNewGame')
        
        fireEvent.click(getByTestId('button-incorrect'))
        
        expect(makeWinnerAlertSpy).toBeCalled()
        expect(chooseNewRandomGameSpy).toBeCalled()
    })
})

/* --------------------------------- */
/* ----- Test help functions ------- */
/* --------------------------------- */

function isPlayersIncludedInGameEvent(textContentCurrentPlayers: string | null, currentPlayers: Player[]): boolean {    
    if (!textContentCurrentPlayers) {
        return false
    }

    return currentPlayers.every((player: Player) => textContentCurrentPlayers.includes(player.name))
}
