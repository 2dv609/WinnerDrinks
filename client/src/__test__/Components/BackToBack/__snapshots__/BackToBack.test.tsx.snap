// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game module should match snapshot 1`] = `
<div
  className="box container"
>
  <div
    className="content"
    data-testid="game-event"
  />
  <h3
    className="title is-6"
  >
    Were both correct?
  </h3>
  <button
    className="button"
    onClick={[Function]}
  >
    They were correct!
  </button>
  <button
    className="button"
    onClick={[Function]}
  >
    They were wrong!
  </button>
</div>
`;
